Add a streaming json parser

nfcampos opened the pull request:
<img width="1728" alt="Screenshot 2023-09-28 at 20 15 01" src="https://github.com/langchain-ai/langchain/assets/56902/ed0644c3-6db7-41b9-9543-e34fce46d3e5">


<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - **Description:** a description of the change, 
  - **Issue:** the issue # it fixes (if applicable),
  - **Dependencies:** any dependencies required for this change,
  - **Tag maintainer:** for a quicker response, tag the relevant maintainer (see below),
  - **Twitter handle:** we announce bigger features on Twitter. If your PR gets announced, and you'd like a mention, we'll gladly shout you out!

Please make sure your PR is passing linting and testing before submitting. Run `make format`, `make lint` and `make test` to check this locally.

See contribution guidelines for more information on how to write/run tests, lint, etc: 
https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md

If you're adding a new integration, please include:
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/extras` directory.

If no one reviews your PR within a few days, please @-mention one of @baskaryan, @eyurtsev, @hwchase17.
 -->

nfcampos committed:
Add a streaming json parser

nfcampos committed:
Implement str one

eyurtsev commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -65,7 +127,7 @@ def parse_json_markdown(json_string: str) -> dict:
 
     # Parse the JSON string into a Python dictionary
-    parsed = json.loads(json_str)
+    parsed = parse_partial_json(json_str)
```
do we need a partial version for the '```' segment as well or does `custom_parser` or another step does stripping of ```?

nfcampos replied:
Not sure I understand the question?

nfcampos committed:
WIP Add tests§

nfcampos committed:
Implement diff

nfcampos committed:
Implement diff

nfcampos committed:
Backwards compat

eyurtsev commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -101,10 +168,54 @@ class SimpleJsonOutputParser(BaseOutputParser[Any]):
+    def _diff(self, prev: Optional[Any], next: Any) -> Any:
+        return jsonpatch.make_patch(prev, next).patch
+
+    def parse(self, text: str) -> Any:
```
Could you leave a comment here to explain why this is not implemented? Could we raise an Exception instead to make sure that this code remains unreachable?

eyurtsev commented on libs/langchain/langchain/schema/output_parser.py:
```diff
@@ -329,6 +334,69 @@ async def atransform(
+            else:
+                acc_gen += chunk_gen
+
+            parsed = self.parse_result([acc_gen])
```
Should this be using aparse_result if it's available?

eyurtsev commented on libs/langchain/langchain/schema/output_parser.py:
```diff
@@ -329,6 +334,69 @@ async def atransform(
+
+    diff: bool = False
+
+    def _diff(self, prev: Optional[T], next: T) -> T:
```
Perhaps document this method?

Do you forsee a situation in which `_diff` is defined but the user passes `diff=False` ?

eyurtsev commented on libs/langchain/tests/unit_tests/output_parsers/test_json.py:
```diff
@@ -183,3 +192,303 @@ def test_parse_json_with_python_dict() -> None:
+    [{"op": "replace", "path": "/audience/1", "value": "So funny"}],
+]
+
+
```
is it worthwhile to test the async path?

eyurtsev commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -65,7 +132,7 @@ def parse_json_markdown(json_string: str) -> dict:
 
     # Parse the JSON string into a Python dictionary
-    parsed = json.loads(json_str)
+    parsed = parser(json_str)
```
Not sure if I'm misreading the regexp, but it seems to require closing ```? Should that requirement be relaxed?

nfcampos replied:
Not sure, but I don't want to change that in this pr

nfcampos committed:
Add async tests and comments

nfcampos committed:
Make test deterministic

nfcampos committed:
Combine with existing json output parsers

nfcampos committed:
Lint

nfcampos committed:
Keep exceptions when not in streaming mode

eyurtsev commented on libs/langchain/langchain/schema/output_parser.py:
```diff
@@ -29,7 +34,7 @@ class BaseLLMOutputParser(Serializable, Generic[T], ABC):
 
     @abstractmethod
-    def parse_result(self, result: List[Generation]) -> T:
+    def parse_result(self, result: List[Generation], *, partial: bool = False) -> T:
```
could we update doc-strings to explain what partial does

eyurtsev commented on libs/langchain/langchain/schema/output_parser.py:
```diff
@@ -329,6 +338,74 @@ async def atransform(
+            else:
+                acc_gen += chunk_gen
+
+            parsed = self.parse_result([acc_gen], partial=True)
```
should this use aparse_result if it's available?

nfcampos replied:
I don't think so because the default implementation of aparse_result is a bit silly. Eg it would launch a thread just to call json.loads

eyurtsev commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -95,13 +163,23 @@ def parse_and_check_json_markdown(text: str, expected_keys: List[str]) -> dict:
         text = text.strip()
         try:
-            return json.loads(text)
+            return parse_json_markdown(text.strip(), parse_partial_json)
```
```suggestion
            return parse_json_markdown(text.strip(), parser=parse_partial_json)
```

eyurtsev commented on libs/langchain/langchain/output_parsers/json.py:
```diff
@@ -38,7 +43,70 @@ def _custom_parser(multiline_string: str) -> str:
+
+
+def parse_json_markdown(
+    json_string: str, parser: Callable[[str], Any] = json.loads
```
```suggestion
    json_string: str, *, parser: Callable[[str], Any] = json.loads
```

eyurtsev commented on libs/langchain/langchain/output_parsers/openai_functions.py:
```diff
@@ -17,7 +23,7 @@ class OutputFunctionsParser(BaseGenerationOutputParser[Any]):
     """Whether to only return the arguments to the function call."""
 
-    def parse_result(self, result: List[Generation]) -> Any:
+    def parse_result(self, result: List[Generation], *, partial: bool = False) -> Any:
```
should this raise on patial = True?

eyurtsev commented on libs/langchain/langchain/output_parsers/openai_functions.py:
```diff
@@ -72,7 +125,7 @@ class JsonKeyOutputFunctionsParser(JsonOutputFunctionsParser):
 
-    def parse_result(self, result: List[Generation]) -> Any:
+    def parse_result(self, result: List[Generation], *, partial: bool = False) -> Any:
         res = super().parse_result(result)
```
propagate partial?

eyurtsev commented on libs/langchain/langchain/output_parsers/openai_functions.py:
```diff
@@ -97,7 +150,7 @@ def validate_schema(cls, values: Dict) -> Dict:
         return values
 
-    def parse_result(self, result: List[Generation]) -> Any:
+    def parse_result(self, result: List[Generation], *, partial: bool = False) -> Any:
```
propagate partial and have underlying implementation raise if it's not supported?

nfcampos committed:
Update json.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

nfcampos committed:
Update json.py

Co-authored-by: Eugene Yurtsev <eyurtsev@gmail.com>

nfcampos committed:
Lint

nfcampos committed:
Remove flawed test

- It is not possible to access properties on classes, only on instances, therefore this test is not something we can implement
